### DISCUSSION TITLE
How do I get OSSL_set_max_threads() to behave nicely with the fips provider?

### DISCUSSION THREAD

**Discussion by slontis** [openssl/openssl#22448] [General Discussion]
OSSL_set_max_threads() was added for ARGON2 support.. It is associated with a libctx which is the first param.

Trying to use this inside a dispatch function inside the fips provider by calling ossl_get_avail_threads() will fail since the libctx is the one associated with the fips provider (not the core libctx).

If argon ever moves to the fips provider it will break in this scenario.
I am trying to do similar code in LMS which breaks when doing fips provider tests with threading..

There is a c_get_libctx() in the fipsprovider but it is not exposed to anything other than the self test callback which also needs the parent libctx..

Is there a nice way I can get to the parent libctx? (Looks like it would need a global function that is both in fips and core..

FIPS
````
OSSL_LIB_CTX *OSSL_LIB_CTX_get_core(OSSL_LIB_CTX *libctx)
{
   if (c_get_libctx == NULL)
       return NULL;
    return c_get_libctx(FIPS_get_core_handle(libctx));
}
````   
CORE
````
OSSL_LIB_CTX *OSSL_LIB_CTX_get_core(OSSL_LIB_CTX *libctx)
{
    return libctx;
}
````

**Comment by slontis** [comment]
Hmm c_get_libctx() returns a OPENSSL_CORE_CTX.. If I cast it and call ossL_get_avail_threads() it segfaults since tdata is garbage,,

**Comment by mattcaswell** [comment]
You cannot use the libctx from core in a provider. It works in default because default is cheating. E.g. consider the case where the layout of the libctx structure changes from one version of OpenSSL to another (as it did between 3.0 and 3.1 IIRC).

Can't we make `ossl_get_avail_threads` available as a core callback?

**Reply by t8m** [reply]
Yeah that would seem to be a logical solution.

**Reply by paulidale** [reply]
A core callback is the only solution here.

**Comment by slontis** [comment]
Had to also add callbacks for thread_start, join and cleanup
